bucket4j/src/main/java/com/linecorp/armeria/server/throttling/bucket4j/BandwidthLimit.java

```diff
@@ -0,0 +1,210 @@
+     *              and defines the speed at which tokens are regenerated in the bucket
+     * @param overdraftLimit defines the maximum overdraft count of tokens which can be held by the bucket,
+     *                       this value must exceed the {@code limit}
+     * @param period the time window, during which the tokens will be regenerated
+     */
+    public static BandwidthLimit of(long limit, long overdraftLimit, Duration period) {
+        return new BandwidthLimit(limit, overdraftLimit, 0L, period);
+    }
+
+    /**
+     * Returns a newly created simple {@link BandwidthLimit}.
+     * Specifies easy limitation of {@code limit} tokens per {@code period} time window.
+     * @param limit the bucket size - defines the maximum count of tokens which can be held by the bucket
+     *              and defines the speed at which tokens are regenerated in the bucket
+     * @param period the time window, during which the tokens will be regenerated
+     */
+    public static BandwidthLimit of(long limit, Duration period) {
+        return new BandwidthLimit(limit, 0L, 0L, period);
+    }
+
+    /**
+     * Returns a newly created {@link BandwidthLimit}. Computes {@code limit}, {@code overdraftLimit},
+     * {@code initialSize} and {@code period} out of a semicolon-separated {@code specification} string
+     * that conforms to the following format,
+     * as per <a href="https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html">RateLimit Header Scheme for HTTP</a>:
+     * <pre>{@code
+     * <limit>;window=<period(in seconds)>[;burst=<overdraftLimit>][;initial=<initialSize>]
+     * }</pre>
+     * All {@code specification} string elements must come in the defined order.
+     * For example:
+     * <ul>
+     *   <li>{@code 100;window=60;burst=1000} ({@code limit}=100, {@code overdraftLimit}=1000,
+     *       {@code initialSize} and {@code period}=60seconds)</li>
+     *   <li>{@code 100;window=60;burst=1000;initial=20} ({@code limit}=100, {@code overdraftLimit}=1000,
+     *       {@code initialSize}=20 and {@code period}=60seconds)</li>
+     *   <li>{@code 5000;window=1} ({@code limit}=5000 and {@code period}=1second)</li>
+     * </ul>
+     *
+     * @param specification the specification used to create a {@link BandwidthLimit}
+     */
+    public static BandwidthLimit of(String specification) {
+        return TokenBucketSpec.parseBandwidthLimit(specification);
+    }
+
+    private final long limit;
+    private final long overdraftLimit;
+    private final long initialSize;
+    private final Duration period;
+
+    BandwidthLimit(long limit, long overdraftLimit, long initialSize, Duration period) {
+        // validate limit
+        if (limit <= 0L) {
+            throw new IllegalArgumentException("Bandwidth limit must be positive. Found: " + limit);
+        }
+        this.limit = limit;
+
+        // validate overdraftLimit
+        if (overdraftLimit > 0L && overdraftLimit <= limit) {
```

Comment:
I think this will not catch when `overdraftLimit` is negative.
We usually just use `checkArgument`, how about using it?
```java
checkArgument(limit > 0, "limit: %s (expected: >0 )", limit);
checkArgument(overdraftLimit == 0 || overdraftLimit > limit,
              "overdraftLimit: %s (expected: > %s )", overdraftLimit, limit);
```

Reply:
Fixed